File: go/internal/registry/registry_test.go
package registry

import (
	"context"
	"errors"
	"strings"
	"sync"
	"sync/atomic"
	"testing"
	"time"
)

func TestKeyFormat(t *testing.T) {
	key, err := GenerateKey()
	if err != nil {
		t.Fatal(err)
	}
	if !ValidFormat(key) {
		t.Fatalf("generated key %q fails its own format check", key)
	}
	other, _ := GenerateKey()
	if key == other {
		t.Error("two generated keys are identical")
	}

	for _, bad := range []string{"", "rlk_short", "xyz_" + strings.Repeat("a", 32), "rlk_" + strings.Repeat("a", 31) + "!", "rlk_" + strings.Repeat("a", 33)} {
		if ValidFormat(bad) {
			t.Errorf("ValidFormat(%q) = true, want false", bad)
		}
	}
	// FIPS 180-2 test vector: SHA-256("abc").
	if got := Hash("abc"); got != "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad" {
		t.Errorf("Hash(\"abc\") = %s, want the SHA-256 test vector", got)
	}
}

func TestParsePlans(t *testing.T) {
	plans, err := ParsePlans(" free=100 , paid=1000 ")
	if err != nil || plans["free"] != 100 || plans["paid"] != 1000 {
		t.Fatalf("got %v, %v", plans, err)
	}
	for _, bad := range []string{"", "free", "free=0", "free=-5", "free=abc"} {
		if _, err := ParsePlans(bad); err == nil {
			t.Errorf("ParsePlans(%q) accepted invalid input", bad)
		}
	}
}

func TestIdentityFor(t *testing.T) {
	plans := Plans{"free": 100, "paid": 1000}
	hash := Hash("rlk_" + strings.Repeat("a", 32))

	tests := []struct {
		name   string
		tenant Tenant
		want   Identity
		ok     bool
	}{
		{"isolated: per-key identity at the plan limit", Tenant{ID: "acme", Plan: "paid", Keys: 10, Mode: ModeIsolated}, Identity{ID: "key:" + hash[:16], Limit: 1000}, true},
		{"pooled: tenant identity at plan limit x slots", Tenant{ID: "acme", Plan: "paid", Keys: 10, Mode: ModePooled}, Identity{ID: "tenant:acme", Limit: 10000}, true},
		{"free pooled with one slot", Tenant{ID: "hobby", Plan: "free", Keys: 1, Mode: ModePooled}, Identity{ID: "tenant:hobby", Limit: 100}, true},
		{"zero slots treated as one", Tenant{ID: "odd", Plan: "free", Keys: 0, Mode: ModePooled}, Identity{ID: "tenant:odd", Limit: 100}, true},
		{"unknown plan is not honored", Tenant{ID: "x", Plan: "gold", Keys: 1, Mode: ModePooled}, Identity{}, false},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, ok := plans.identityFor(tt.tenant, hash)
			if got != tt.want || ok != tt.ok {
				t.Errorf("got %+v %v, want %+v %v", got, ok, tt.want, tt.ok)
			}
		})
	}
}

// fakeLookup serves registry records and counts round-trips.
type fakeLookup struct {
	mu      sync.Mutex
	records map[string][]string
	fail    bool
	delay   time.Duration
	calls   atomic.Int64
}

func (f *fakeLookup) ResolveKey(_ context.Context, hash string) ([]string, error) {
	f.calls.Add(1)
	time.Sleep(f.delay)
	f.mu.Lock()
	defer f.mu.Unlock()
	if f.fail {
		return nil, errors.New("redis down")
	}
	return f.records[hash], nil
}

func newResolver(t *testing.T, ttl time.Duration, size int) (*Resolver, *fakeLookup, string) {
	t.Helper()
	key, _ := GenerateKey()
	lookup := &fakeLookup{records: map[string][]string{
		Hash(key): {"acme", "paid", "3", ModePooled},
	}}
	return NewResolver(lookup, Plans{"free": 100, "paid": 1000}, ttl, size), lookup, key
}

func TestResolverCachesHitsAndMisses(t *testing.T) {
	r, lookup, key := newResolver(t, time.Minute, 100)
	ctx := context.Background()

	for i := 0; i < 3; i++ {
		if id, ok := r.Resolve(ctx, key); !ok || id != (Identity{ID: "tenant:acme", Limit: 3000}) {
			t.Fatalf("got %+v %v", id, ok)
		}
	}
	unknown, _ := GenerateKey()
	for i := 0; i < 3; i++ {
		if _, ok := r.Resolve(ctx, unknown); ok {
			t.Fatal("unregistered key resolved")
		}
	}
	if calls := lookup.calls.Load(); calls != 2 {
		t.Errorf("lookups = %d, want 2 (one per key; negative results cached too)", calls)
	}
}

func TestResolverSkipsLookupForMalformedKeys(t *testing.T) {
	r, lookup, _ := newResolver(t, time.Minute, 100)
	if _, ok := r.Resolve(context.Background(), "made-up-key"); ok {
		t.Fatal("malformed key resolved")
	}
	if calls := lookup.calls.Load(); calls != 0 {
		t.Errorf("lookups = %d, want 0 for a malformed key", calls)
	}
}

func TestResolverDoesNotCacheFailures(t *testing.T) {
	r, lookup, key := newResolver(t, time.Minute, 100)
	ctx := context.Background()

	lookup.fail = true
	if _, ok := r.Resolve(ctx, key); ok {
		t.Fatal("resolved during an outage")
	}
	lookup.fail = false
	if _, ok := r.Resolve(ctx, key); !ok {
		t.Error("a failed lookup was cached: the key stayed unresolved after recovery")
	}
}

func TestResolverExpiresEntries(t *testing.T) {
	r, lookup, key := newResolver(t, 20*time.Millisecond, 100)
	ctx := context.Background()

	r.Resolve(ctx, key)
	lookup.mu.Lock()
	delete(lookup.records, Hash(key)) // revoked
	lookup.mu.Unlock()

	if _, ok := r.Resolve(ctx, key); !ok {
		t.Fatal("expected the cached identity before the TTL elapses")
	}
	time.Sleep(30 * time.Millisecond)
	if _, ok := r.Resolve(ctx, key); ok {
		t.Error("revoked key still honored after the TTL")
	}
}

func TestResolverBoundsCacheSize(t *testing.T) {
	r, _, _ := newResolver(t, time.Minute, 10)
	for i := 0; i < 50; i++ {
		k, _ := GenerateKey()
		r.Resolve(context.Background(), k)
	}
	if n := len(r.cache); n != 10 {
		t.Errorf("cache holds %d entries, want it capped at 10", n)
	}
}

func TestResolverDeduplicatesConcurrentMisses(t *testing.T) {
	r, lookup, key := newResolver(t, time.Minute, 100)
	lookup.delay = 20 * time.Millisecond

	var wg sync.WaitGroup
	for i := 0; i < 50; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			if _, ok := r.Resolve(context.Background(), key); !ok {
				t.Error("concurrent resolve failed")
			}
		}()
	}
	wg.Wait()

	if calls := lookup.calls.Load(); calls != 1 {
		t.Errorf("lookups = %d, want 1 for 50 concurrent misses on one key", calls)
	}
}

File: node/test/unit/registry.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { validFormat, hashKey, parsePlans, identityFor, createResolver, ModeIsolated, ModePooled } from '../../lib/registry.js';

const KEY = `rlk_${'a'.repeat(32)}`;
const OTHER = `rlk_${'b'.repeat(32)}`;
const plans = parsePlans('free=100,paid=1000');

test('key format', () => {
  assert.equal(validFormat(KEY), true);
  for (const bad of ['', 'rlk_short', `xyz_${'a'.repeat(32)}`, `rlk_${'a'.repeat(31)}!`, `rlk_${'a'.repeat(33)}`]) {
    assert.equal(validFormat(bad), false, bad);
  }
});

test('hash is SHA-256 (FIPS 180-2 test vector)', () => {
  assert.equal(hashKey('abc'), 'ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad');
});

test('parsePlans accepts tiers and rejects garbage', () => {
  assert.deepEqual([...parsePlans(' free=100 , paid=1000 ')], [['free', 100], ['paid', 1000]]);
  for (const bad of ['', 'free', 'free=0', 'free=-5', 'free=abc']) {
    assert.throws(() => parsePlans(bad), undefined, bad);
  }
});

test('identityFor: isolated, pooled, free, unknown plan', () => {
  const hash = hashKey(KEY);
  assert.deepEqual(identityFor(plans, { id: 'acme', plan: 'paid', keys: 10, mode: ModeIsolated }, hash), { id: `key:${hash.slice(0, 16)}`, limit: 1000 });
  assert.deepEqual(identityFor(plans, { id: 'acme', plan: 'paid', keys: 10, mode: ModePooled }, hash), { id: 'tenant:acme', limit: 10000 });
  assert.deepEqual(identityFor(plans, { id: 'hobby', plan: 'free', keys: 0, mode: ModePooled }, hash), { id: 'tenant:hobby', limit: 100 });
  assert.equal(identityFor(plans, { id: 'x', plan: 'gold', keys: 1, mode: ModePooled }, hash), null);
});

function lookupWith(records, { fail = false, delayMs = 0 } = {}) {
  const lookup = async (hash) => {
    lookup.calls += 1;
    if (delayMs) await new Promise((r) => setTimeout(r, delayMs));
    if (lookup.fail) throw new Error('redis down');
    return records.get(hash) ?? [];
  };
  lookup.calls = 0;
  lookup.fail = fail;
  return lookup;
}

test('caches registered and unregistered results', async () => {
  const lookup = lookupWith(new Map([[hashKey(KEY), ['acme', 'paid', '3', ModePooled]]]));
  const resolve = createResolver({ lookup, plans, ttlMs: 60_000, size: 100 });
  for (let i = 0; i < 3; i++) assert.deepEqual(await resolve(KEY), { id: 'tenant:acme', limit: 3000 });
  for (let i = 0; i < 3; i++) assert.equal(await resolve(OTHER), null);
  assert.equal(lookup.calls, 2);
});

test('malformed keys never reach Redis', async () => {
  const lookup = lookupWith(new Map());
  const resolve = createResolver({ lookup, plans, ttlMs: 60_000, size: 100 });
  assert.equal(await resolve('made-up-key'), null);
  assert.equal(lookup.calls, 0);
});

test('failed lookups are not cached', async () => {
  const lookup = lookupWith(new Map([[hashKey(KEY), ['acme', 'free', '1', ModePooled]]]), { fail: true });
  const resolve = createResolver({ lookup, plans, ttlMs: 60_000, size: 100 });
  assert.equal(await resolve(KEY), null);
  lookup.fail = false;
  assert.deepEqual(await resolve(KEY), { id: 'tenant:acme', limit: 100 });
});

test('entries expire after the TTL (revocation)', async () => {
  const records = new Map([[hashKey(KEY), ['acme', 'free', '1', ModePooled]]]);
  const resolve = createResolver({ lookup: lookupWith(records), plans, ttlMs: 20, size: 100 });
  await resolve(KEY);
  records.delete(hashKey(KEY));
  assert.notEqual(await resolve(KEY), null, 'still cached within the TTL');
  await new Promise((r) => setTimeout(r, 30));
  assert.equal(await resolve(KEY), null, 'revoked key honored after the TTL');
});

test('cache size is bounded', async () => {
  const lookup = lookupWith(new Map());
  const resolve = createResolver({ lookup, plans, ttlMs: 60_000, size: 10 });
  for (let i = 0; i < 50; i++) await resolve(`rlk_${String(i).padStart(32, '0')}`);
  lookup.calls = 0;
  await resolve(`rlk_${String(0).padStart(32, '0')}`); // oldest: evicted
  await resolve(`rlk_${String(49).padStart(32, '0')}`); // newest: cached
  assert.equal(lookup.calls, 1);
});

test('concurrent misses on one key share a single lookup', async () => {
  const lookup = lookupWith(new Map([[hashKey(KEY), ['acme', 'free', '1', ModePooled]]]), { delayMs: 20 });
  const resolve = createResolver({ lookup, plans, ttlMs: 60_000, size: 100 });
  const results = await Promise.all(Array.from({ length: 50 }, () => resolve(KEY)));
  assert.ok(results.every((r) => r?.id === 'tenant:acme'));
  assert.equal(lookup.calls, 1);
});
